# Week 15 : Case Study 10 - Bimal H M

## Probability

### Problem Statement : 
<b>To make a suitable machine learning algorithm to predict if the mushroom is edible or poisonous (e or p) using the given dataset. (Along with other ML algorithms, Naïve Bayes’ Classifier should be applied) Also, if some data pre-processing is necessary do that as well.<b>

## STEP 1 : Importing all Necessary Libraries in to Notebook

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## STEP 2 : Read the dataset to the python environment

In [2]:
data = pd.read_csv('mushrooms.csv')

## STEP 3 : Checking Basic Details

#### A) Displaying First few Rows

In [3]:
data.head()

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0     p         x           s         n       t    p               f   
1     e         x           s         y       t    a               f   
2     e         b           s         w       t    l               f   
3     p         x           y         w       t    p               f   
4     e         x           s         g       f    n               f   

  gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0            c         n          k  ...                        s   
1            c         b          k  ...                        s   
2            c         b          n  ...                        s   
3            c         n          n  ...                        s   
4            w         b          k  ...                        s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

#### B) Displaying Dataset Info

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8124 entries, 0 to 8123
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   class                     8124 non-null   object
 1   cap-shape                 8124 non-null   object
 2   cap-surface               8124 non-null   object
 3   cap-color                 8124 non-null   object
 4   bruises                   8124 non-null   object
 5   odor                      8124 non-null   object
 6   gill-attachment           8124 non-null   object
 7   gill-spacing              8124 non-null   object
 8   gill-size                 8124 non-null   object
 9   gill-color                8124 non-null   object
 10  stalk-shape               8124 non-null   object
 11  stalk-root                8124 non-null   object
 12  stalk-surface-above-ring  8124 non-null   object
 13  stalk-surface-below-ring  8124 non-null   object
 14  stalk-color-above-ring  

#### C) Checking for summary statistics

In [5]:
data.describe()

class cap-shape cap-surface cap-color bruises  odor gill-attachment  \
count   8124      8124        8124      8124    8124  8124            8124   
unique     2         6           4        10       2     9               2   
top        e         x           y         n       f     n               f   
freq    4208      3656        3244      2284    4748  3528            7914   

       gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
count          8124      8124       8124  ...                     8124   
unique            2         2         12  ...                        4   
top               c         b          b  ...                        s   
freq           6812      5612       1728  ...                     4936   

       stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
count                    8124                   8124      8124       8124   
unique                      9                      9         1          4   
top                         w                      w         p          w   
freq                     4464                   4384      8124       7924   

       ring-number ring-type spore-print-color population habitat  
count         8124      8124              8124       8124    8124  
unique           3         5                 9          6       7  
top              o         p                 w          v       d  
freq          7488      3968              2388       4040    3148  

[4 rows x 23 columns]

###### From dataset info  it is clear that all elements are of object type and there is no null value

## STEP 4 : Pre-processing steps

### A) Label Encoding

all elements in dataset are object type so we should do a label encoding to convert them to int datatypes

In [6]:
data.columns

Index(['class', 'cap-shape', 'cap-surface', 'cap-color', 'bruises', 'odor',
       'gill-attachment', 'gill-spacing', 'gill-size', 'gill-color',
       'stalk-shape', 'stalk-root', 'stalk-surface-above-ring',
       'stalk-surface-below-ring', 'stalk-color-above-ring',
       'stalk-color-below-ring', 'veil-type', 'veil-color', 'ring-number',
       'ring-type', 'spore-print-color', 'population', 'habitat'],
      dtype='object')

In [7]:
from sklearn.preprocessing import LabelEncoder

LE = LabelEncoder()
for i in data.columns:
    data[i] = LE.fit_transform(data[i])

In [8]:
data.head()

class  cap-shape  cap-surface  cap-color  bruises  odor  gill-attachment  \
0      1          5            2          4        1     6                1   
1      0          5            2          9        1     0                1   
2      0          0            2          8        1     3                1   
3      1          5            3          8        1     6                1   
4      0          5            2          3        0     5                1   

   gill-spacing  gill-size  gill-color  ...  stalk-surface-below-ring  \
0             0          1           4  ...                         2   
1             0          0           4  ...                         2   
2             0          0           5  ...                         2   
3             0          1           5  ...                         2   
4             1          0           4  ...                         2   

   stalk-color-above-ring  stalk-color-below-ring  veil-type  veil-color  \
0                       7                       7          0           2   
1                       7                       7          0           2   
2                       7                       7          0           2   
3                       7                       7          0           2   
4                       7                       7          0           2   

   ring-number  ring-type  spore-print-color  population  habitat  
0            1          4                  2           3        5  
1            1          4                  3           2        1  
2            1          4                  3           2        3  
3            1          4                  2           3        5  
4            1          0                  3           0        1  

[5 rows x 23 columns]

### B) Checking for Unique values and Dimensionality Reduction

In [9]:
for i in data.columns:
    print(i,'-',len(data[i].unique()))

class - 2
cap-shape - 6
cap-surface - 4
cap-color - 10
bruises - 2
odor - 9
gill-attachment - 2
gill-spacing - 2
gill-size - 2
gill-color - 12
stalk-shape - 2
stalk-root - 5
stalk-surface-above-ring - 4
stalk-surface-below-ring - 4
stalk-color-above-ring - 9
stalk-color-below-ring - 9
veil-type - 1
veil-color - 4
ring-number - 3
ring-type - 5
spore-print-color - 9
population - 6
habitat - 7


###### It is clear that there is no Missing values present in dataset  and outliers doesnot hold enough significance since whole dataset is of object type. and from uniquevalue count it is also clear that "veil-type" doesnot have enough significence in the dataset(as only one value).

## STEP 5 : Classification Models

In [10]:
import warnings
warnings.filterwarnings('ignore')

#### Preparing Test and Train Set

In [11]:
y = data['class']
x = data.drop(['class','veil-type'],axis = 1)

In [12]:
#Parting dataset in to Test and Train and importing necessery libraries

from sklearn.model_selection import train_test_split

x_train,x_test,y_train,y_test = train_test_split(x,y,random_state=42,test_size=0.2)

from sklearn.metrics import accuracy_score,confusion_matrix,classification_report

### A) Logistic Regression

In [13]:
from sklearn.linear_model import LogisticRegression 

LR = LogisticRegression()
LR.fit(x_train,y_train)
y_pred = LR.predict(x_test)

print('Accuracy score         = ',round(accuracy_score(y_test,y_pred),3),'\n\nConfusion matrix       = \n\n',confusion_matrix(y_test,y_pred),'\n\nClassification report  = \n\n',classification_report(y_test,y_pred))

Accuracy score         =  0.946 

Confusion matrix       = 

 [[798  45]
 [ 42 740]] 

Classification report  = 

               precision    recall  f1-score   support

           0       0.95      0.95      0.95       843
           1       0.94      0.95      0.94       782

    accuracy                           0.95      1625
   macro avg       0.95      0.95      0.95      1625
weighted avg       0.95      0.95      0.95      1625



### B) SVM - Support Vector Machine

In [14]:
from sklearn.svm import SVC

svm = SVC(kernel='rbf')
svm.fit(x_train,y_train)
y_pred = svm.predict(x_test)

print('Accuracy score         = ',round(accuracy_score(y_test,y_pred),3),'\n\nConfusion matrix       = \n\n',confusion_matrix(y_test,y_pred),'\n\nClassification report  = \n\n',classification_report(y_test,y_pred))

Accuracy score         =  0.994 

Confusion matrix       = 

 [[842   1]
 [  9 773]] 

Classification report  = 

               precision    recall  f1-score   support

           0       0.99      1.00      0.99       843
           1       1.00      0.99      0.99       782

    accuracy                           0.99      1625
   macro avg       0.99      0.99      0.99      1625
weighted avg       0.99      0.99      0.99      1625



### C) Decision Tree

In [15]:
from sklearn.tree import DecisionTreeClassifier

DT = DecisionTreeClassifier()
DT.fit(x_train,y_train)
y_pred = DT.predict(x_test)

print('Accuracy score         = ',round(accuracy_score(y_test,y_pred),3),'\n\nConfusion matrix       = \n\n',confusion_matrix(y_test,y_pred),'\n\nClassification report  = \n\n',classification_report(y_test,y_pred))

Accuracy score         =  1.0 

Confusion matrix       = 

 [[843   0]
 [  0 782]] 

Classification report  = 

               precision    recall  f1-score   support

           0       1.00      1.00      1.00       843
           1       1.00      1.00      1.00       782

    accuracy                           1.00      1625
   macro avg       1.00      1.00      1.00      1625
weighted avg       1.00      1.00      1.00      1625



### D) Random Forest 

In [16]:
from sklearn.ensemble import RandomForestClassifier

RF = RandomForestClassifier()
RF.fit(x_train,y_train)
y_pred = RF.predict(x_test)

print('Accuracy score         = ',round(accuracy_score(y_test,y_pred),3),'\n\nConfusion matrix       = \n\n',confusion_matrix(y_test,y_pred),'\n\nClassification report  = \n\n',classification_report(y_test,y_pred))

Accuracy score         =  1.0 

Confusion matrix       = 

 [[843   0]
 [  0 782]] 

Classification report  = 

               precision    recall  f1-score   support

           0       1.00      1.00      1.00       843
           1       1.00      1.00      1.00       782

    accuracy                           1.00      1625
   macro avg       1.00      1.00      1.00      1625
weighted avg       1.00      1.00      1.00      1625



### E) Probability Model - Naive Bayes 

#### 1 : Gaussian NB

In [17]:
from sklearn.naive_bayes import GaussianNB

GNB = GaussianNB()
GNB.fit(x_train,y_train)
y_pred = GNB.predict(x_test)

print('Accuracy score         = ',round(accuracy_score(y_test,y_pred),3),'\n\nConfusion matrix       = \n\n',confusion_matrix(y_test,y_pred),'\n\nClassification report  = \n\n',classification_report(y_test,y_pred))

Accuracy score         =  0.922 

Confusion matrix       = 

 [[771  72]
 [ 55 727]] 

Classification report  = 

               precision    recall  f1-score   support

           0       0.93      0.91      0.92       843
           1       0.91      0.93      0.92       782

    accuracy                           0.92      1625
   macro avg       0.92      0.92      0.92      1625
weighted avg       0.92      0.92      0.92      1625



#### 2 : Bernoulli NB

In [18]:
from sklearn.naive_bayes import BernoulliNB

BNB = BernoulliNB()
BNB.fit(x_train,y_train)
y_pred=BNB.predict(x_test)

print('Accuracy score         = ',round(accuracy_score(y_test,y_pred),3),'\n\nConfusion matrix       = \n\n',confusion_matrix(y_test,y_pred),'\n\nClassification report  = \n\n',classification_report(y_test,y_pred))

Accuracy score         =  0.842 

Confusion matrix       = 

 [[797  46]
 [211 571]] 

Classification report  = 

               precision    recall  f1-score   support

           0       0.79      0.95      0.86       843
           1       0.93      0.73      0.82       782

    accuracy                           0.84      1625
   macro avg       0.86      0.84      0.84      1625
weighted avg       0.86      0.84      0.84      1625



##### From these Supervised ML Models it is clear that Decision Tree, Random Forrest show an accuracy of 1 and Gausiian NB is the best probabilistic models with 0.92 accuracy.